package/MDAnalysis/analysis/rdf.py

```diff
@@ -187,24 +187,32 @@ def _prepare(self):
 
         # Need to know average volume
         self.volume = 0.0
+        # Set the max range to filter the search radius
+        self._maxrange = self.rdf_settings['range'][1]
 
         # Allocate a results array which we will reuse
-        self._result = np.zeros((len(self.g1), len(self.g2)), dtype=np.float64)
+        self._result = np.ones((len(self.g1), len(self.g2)), dtype=np.float64)
+        # Pre-assign the largest values to avoid 
+        # counting the atoms farther than the maxrange
+        self._result *= self._maxrange + 1
+
         # If provided exclusions, create a mask of _result which
         # lets us take these out
         if self._exclusion_block is not None:
             self._exclusion_mask = blocks_of(self._result,
                                              *self._exclusion_block)
-            self._maxrange = self.rdf_settings['range'][1] + 1.0
         else:
             self._exclusion_mask = None
 
     def _single_frame(self):
-        distances.distance_array(self.g1.positions, self.g2.positions,
-                                 box=self.u.dimensions, result=self._result)
+        pairs, dist = distances.capped_distance(self.g1.positions,
+                                                self.g2.positions,
+                                                self._maxrange,
+                                                box=self.u.dimensions)
+        self._result[tuple(zip(*pairs))] = dist
```

Comment:
This looks ugly and slow, but I see that for the exclusion mask feature you need to have the full matrix of distances.  I think it might be better to implement the exclusion mask in a different way to avoid having to have this full matrix representation.  

Because this full matrix scales O(N^2) we should really be avoiding it anyway.  I'll have a think to see if I can come up with a quick way to do exclusion_matrix.   You should be able to go through the indices, check if they're inside an excluded block, and if so ignore that distance.